### DISCUSSION TITLE
A way to add files in the resulting image

### DISCUSSION THREAD

**Discussion by giovannicandido** [paketo-buildpacks/feedback#196] [General]
I'm using spring boot  plugin to build images, which in turn use some build pack. This is the list of buildpacks used:

 paketo-buildpacks/ca-certificates       3.6.1
    [creator]     paketo-buildpacks/bellsoft-liberica     10.2.0
    [creator]     paketo-buildpacks/syft                  1.28.0
    [creator]     paketo-buildpacks/executable-jar        6.6.3
    [creator]     paketo-buildpacks/dist-zip              5.5.2
    [creator]     paketo-buildpacks/spring-boot           5.23.0
    [creator]     paketo-buildpacks/environment-variables 4.5.1

When using the command line to build instead of the plugin it uses the following build pack:

paketo-buildpacks/ca-certificates       3.6.1
[detector] paketo-buildpacks/bellsoft-liberica     10.2.0
[detector] paketo-buildpacks/syft                  1.28.0
[detector] paketo-buildpacks/gradle                7.1.0
[detector] paketo-buildpacks/executable-jar        6.6.3
[detector] paketo-buildpacks/apache-tomcat         7.13.0
[detector] paketo-buildpacks/apache-tomee          1.6.1
[detector] paketo-buildpacks/liberty               3.6.0
[detector] paketo-buildpacks/dist-zip              5.5.2
[detector] paketo-buildpacks/spring-boot           5.23.0
[detector] paketo-buildpacks/environment-variables 4.5.1

When I use the env variable BP_INCLUDE_FILES=newrelic/* as mentioned in the documentation. The file is not added using the plugin, but is added using the command line. Investigation shows that the buildpack which is participating and copying the artifacts is packeto-buildpackes/gradle

Other buildpack which claims to add files using this env variable is paketo-buildpacks/source-removal. But as a user I find this quite confusing and I spend 2 days figuring out why the file is not being added.

Should this behavior be independent of buildpacks added? I think this should be a core feature.

**Comment by giovannicandido** [comment]
Another comment: I'm trying to solve the issue with spring boot plugin by adding the source-removal build pack to the build. But I'm getting this error:

Pulling buildpack image 'docker.io/paketo-buildpacks/source-removal:latest' ..................................................

> Task :bootBuildImage FAILED

I cannot pull the source-removal image from any docker registry... tryed using:

docker pull /paketo-buildpacks/source-removal
docker pull /paketo-buildpacks/source-removal:v0.2.0
docker pull /paketo-buildpacks/source-removal:0.2.0
docker pull ghcr.io/paketo-buildpacks/source-removal

How to add the source-removal buildpack?

I don't think adding the gradle buildpack is the correct answer because spring boot plugin is able to generate the image without using it to build the app.

**Comment by giovannicandido** [comment]
Second note: adding the gradle buildpack but is not picked by the plugin command. I think it is explicity disabled by the plugin

**Comment by dmikusa** [comment]
Things are different from the command line and from Spring Boot's plugin. The Spring Boot plugin will compile your application *locally* and then build the image from your executable JAR.

By default, `pack build` will send your source code into the container, build the JAR in the container, and then build the app image.

When you try to tell it to retain files and you use the Spring Boot plugin, it doesn't work because those files are already gone (or at the least, in a different folder structure because they are in your JAR).

What I would suggest is to put everything you require into your JAR file. This will ensure that it ends up in the resulting container image without any additional configuration. Everything in the JAR ends up in the image, regardless of how you build.

If you cannot do that, then a couple of notes:

- You need to build with `pack`. You cannot make the Spring Boot Plugin build from source inside the container.
- Do not use the source removal buildpack. If you're building from source (with pack) the Maven/Gradle plugins have direct support for `BP_EXCLUDE_FILES` and `BP_INCLUDE_FILES` so you just set those and they'll retain the files. 

Hope that helps!

**Reply by giovannicandido** [reply]
Thanks for the clarification. I did try to embed the agent in the jar with no sucess, that is why I need to copy it.
I will use the pack command, this will modify my CI a little bit as I run tests and building in separeted stages from docker build so it will be just a waste to rebuild the image again in the docker phase, but I will figureout the best possible approach.

Any thoughts on making this a core feature indepent of the build pack used, so it could work on any case, building from source or building locally then adding the files to the final image?

**Reply by giovannicandido** [reply]
The documentation on site makes the impression that it works on any case, as if the builder resposibility not the builderpack. If not implemented as core, I suggest to clarify this behaviour in doc.

**Reply by dmikusa** [reply]
I'm not totally sure what you mean by making it a core feature. With cloud-native buildpacks, there are very few core features, just what is built into pack (or Spring Boot plugin) and the buildpack lifecycle. Most functionality comes from the buildpacks, like we have implemented now. If you want to suggest a change for how the core works, you'd want to ask over on https://github.com/buildpacks/pack or the Spring Boot Github.

I agree, it is a little confusing how you can build from both source or pre-compiled assets. The additional complexity makes for more flexibility though. I know Spring Boot uses pre-compiled assets for the reason you mentioned, it fits better into CI, among others.

**Reply by giovannicandido** [reply]
I like the flexibility of the pre-compiled vs builder. In dockerfile we can direct alter the layers of to resulting image, and I understand why this is not the case here, but  maybe a buildpack that is trigged in the final image builder and gives the possibility to add/remove using docker COPY API files can work on any case, I'm not sure if we need a core feature or just a specific lifecycle (I'm new to buildpacks).
This open other possibilities as well, the core reason is to keep using buildpacks for the easy of use and best pratices but tweak a little bit the resulting image without having to create your own buidlpack which is not a trivial task to do. I think this covers more use cases.

Anyway thanks for the clarification it solves the question.

**Reply by dmikusa** [reply]
> tweak a little bit the resulting image without having to create your own buidlpack which is not a trivial task to do

If you're thinking that all buildpacks need to be like Paketo buildpacks, that's not the right picture. Paketo buildpacks are on the more complicated side of things. We optimize for a very wide range of use cases and it makes the buildpacks more complicated.

If you've not done so, I would encourage you to try writing a buildpack. I hope that folks find it accessible. It minimally requires two shell scripts (or other executables). There's a great tutorial over here: https://buildpacks.io/docs/buildpack-author-guide/create-buildpack/
